File: Data Structures & Algorithms/subsets/submission-3.cs
public class Solution {
    public List<List<int>> Subsets(int[] nums) {
        List<List<int>> res = new();
        List<int> ls = new();
        backtrack(0, nums, res, ls);
        return res;
    }

    public void backtrack(int i, int []nums, List<List<int>> res, List<int> ls){
        if(i==nums.Length){
            res.Add(new List<int>(ls));
            return;
        }
        ls.Add(nums[i]);
        backtrack(i+1, nums, res, ls);
        ls.RemoveAt(ls.Count()-1);
        backtrack(i+1, nums, res, ls);
    }
}

File: Data Structures & Algorithms/subsets-ii/submission-3.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        Set<List<Integer>> res = new HashSet<>();
        backtrack(0, nums, new ArrayList<>(), res);
        return new ArrayList(res);
    }

    public void backtrack(int i, int[] nums, List<Integer> ls, Set<List<Integer>> res){
        if(i==nums.length){
            res.add(new ArrayList<>(ls));
            return;
        }
        ls.add(nums[i]);
        backtrack(i+1, nums, ls, res);
        ls.remove(ls.size()-1);
        backtrack(i+1, nums, ls, res);
    }
}
